demo/primitives/src/transaction.rs

```diff
@@ -0,0 +1,428 @@
+			InternalFunctionId::StakingSetSessionsPerEra =>
+				Proposal::StakingSetSessionsPerEra(Slicable::decode(input)?),
+			InternalFunctionId::StakingSetBondingDuration =>
+				Proposal::StakingSetBondingDuration(Slicable::decode(input)?),
+			InternalFunctionId::StakingSetValidatorCount =>
+				Proposal::StakingSetValidatorCount(Slicable::decode(input)?),
+			InternalFunctionId::StakingForceNewEra => Proposal::StakingForceNewEra,
+			InternalFunctionId::DemocracyCancelReferendum => Proposal::DemocracyCancelReferendum(Slicable::decode(input)?),
+		};
+
+		Some(function)
+	}
+
+	fn encode(&self) -> Vec<u8> {
+		let mut v = Vec::new();
+		match *self {
+			Proposal::SystemSetCode(ref data) => {
+				(InternalFunctionId::SystemSetCode as u8).using_encoded(|s| v.extend(s));
+				data.using_encoded(|s| v.extend(s));
+			}
+			Proposal::SessionSetLength(ref data) => {
+				(InternalFunctionId::SessionSetLength as u8).using_encoded(|s| v.extend(s));
+				data.using_encoded(|s| v.extend(s));
+			}
+			Proposal::SessionForceNewSession => {
+				(InternalFunctionId::SessionForceNewSession as u8).using_encoded(|s| v.extend(s));
+			}
+			Proposal::StakingSetSessionsPerEra(ref data) => {
+				(InternalFunctionId::StakingSetSessionsPerEra as u8).using_encoded(|s| v.extend(s));
+				data.using_encoded(|s| v.extend(s));
+			}
+			Proposal::StakingSetBondingDuration(ref data) => {
+				(InternalFunctionId::StakingSetBondingDuration as u8).using_encoded(|s| v.extend(s));
+				data.using_encoded(|s| v.extend(s));
+			}
+			Proposal::StakingSetValidatorCount(ref data) => {
+				(InternalFunctionId::StakingSetValidatorCount as u8).using_encoded(|s| v.extend(s));
+				data.using_encoded(|s| v.extend(s));
+			}
+			Proposal::StakingForceNewEra => {
+				(InternalFunctionId::StakingForceNewEra as u8).using_encoded(|s| v.extend(s));
+			}
+			Proposal::DemocracyCancelReferendum(ref data) => {
+				(InternalFunctionId::DemocracyCancelReferendum as u8).using_encoded(|s| v.extend(s));
+				data.using_encoded(|s| v.extend(s));
+			}
+		}
+
+		v
+	}
+}
+
+/// Public functions that can be dispatched to.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize, Debug))]
+#[repr(u8)]
+enum FunctionId {
+	TimestampSet = 0x00,
+
+	SessionSetKey = 0x10,
+
+	StakingStake = 0x20,
+	StakingUnstake = 0x21,
+	StakingTransfer = 0x22,
+
+	CouncilVotePropose = 0x30,
+	CouncilVoteVote = 0x31,
+	CouncilVoteVeto = 0x32,
+
+	CouncilSetApprovals = 0x40,
+	CouncilReapInactiveVoter = 0x41,
+	CouncilRetractVoter = 0x42,
+	CouncilSubmitCandidacy = 0x43,
+	CouncilPresentWinner = 0x44,
+
+	DemocracyPropose = 0x50,
+	DemocracySecond = 0x51,
+	DemocracyVote = 0x52,
+}
+
+impl FunctionId {
+	/// Derive `Some` value from a `u8`, or `None` if it's invalid.
+	fn from_u8(value: u8) -> Option<FunctionId> {
+		use self::*;
+		let functions = [FunctionId::StakingStake, FunctionId::StakingUnstake,
```

Comment:
Should not it contain all enum values?

Reply:
this really should be a simple match. sure, it's a bit more verbose, but matches have to be exhaustive and we wouldn't have to worry about this breaking

Reply:
it'll soon not matter.